File: 15.3sum/solution.ts
function threeSum(nums: number[]): number[][] {
    const sorted = nums.sort((a, b) => a - b);
    const indiciesByNumber = new Map<number, number[]>();
    for (let i = 0; i < sorted.length; i++) {
        const indicies: number[] = indiciesByNumber.get(sorted[i]) || [];
        indicies.push(i);
        indiciesByNumber.set(sorted[i], indicies);
    }

    const results: number[][] = [];
    const checkedOnI = new Set<number>();
    for (let i = 0; i < sorted.length - 2; i++) {
        // Skips if sorted[i] is already checked.
        if (checkedOnI.has(sorted[i])) {
            continue;
        }
        checkedOnI.add(sorted[i]);

        const checkedOnJ = new Set<number>();
        for (let j = i + 1; j < sorted.length - 1; j++) {
            // Skips if sorted[j] is already checked.
            if (checkedOnJ.has(sorted[j])) {
                continue;
            }
            checkedOnJ.add(sorted[j]);

            const rest = 0 - sorted[i] - sorted[j];
            if (!indiciesByNumber.has(rest)) {
                continue;
            }
            const indicies = indiciesByNumber.get(rest)!;
            for (let k = 0; k < indicies.length; k++) {
                if (indicies[k] <= j) {
                    continue;
                }
                results.push([
                    sorted[i],
                    sorted[j],
                    sorted[indicies[k]],
                ]);
                break;
            }
        }
    }
    return results;
};

File: 15.3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        
        results = []
        n = len(nums)
        i = 0
        for i in range(n):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            if nums[i] > 0:
                break
            
            j = i + 1
            k = n - 1
            while j < k:
                calc = nums[i] + nums[j] + nums[k]
                if calc == 0:
                    results.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j - 1] == nums[j]:
                        j += 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
                    continue
                if calc < 0:
                    j += 1
                    while j < k and nums[j - 1] == nums[j]:
                        j += 1
                else:
                    k -= 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
        return results